### DISCUSSION TITLE
[3주차] 트랜스포트 계층 3.1~4

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#11] [Q&A]
3주차 트랜스포트 계층 3.1~4에 해당하는 질의응답입니다.

**Comment by H-sooyeon** [comment]
## Q1
Multiplexing과 Demultiplexing의 개념에 대한 간단한 설명과  TCP와 UDP에서의 Multiplexing, Demultiplexing의 차이는 무엇인가? 

## Q2 
DNS가 UDP를 사용하는 것 처럼, 일반적으로 네트워크 개발자는 TCP보다 UDP를 선호한다. 그 이유는 무엇인가?

## Q3
TCP 프로토콜에서 rdt1.0, 2.0, 3.0 각각에 대해 차이점을 기준으로 설명하라.

**Reply by Hanjaemo** [reply]
> ### Q1 
> Multiplexing과 Demultiplexing의 개념에 대한 간단한 설명과 TCP와 UDP에서의 Multiplexing, Demultiplexing의 차이는 무엇인가?

하나의 호스트에는 여러 프로세스가 존재합니다. 이들 중 올바른 프로세스(소켓)가 수신할 수 있도록 데이터를 전달하는 작업을 **Demultiplexing**(역다중화)이라고 합니다.

반대로 여러 소켓으로부터 데이터를 모으고 각 데이터를 헤더로 캡슐화하여 네트워크 계층으로 전달하는 작업을 **Multiplexing**(다중화)이라고 합니다.

그리고 이 다중화/역다중화는 TCP와 UDP에서 각각 다르게 동작합니다.

UDP는 비연결형 프로토콜, 즉 소켓과 소켓 사이에 1:1 연결을 맺지 않습니다. 이 때문에 동일한 소켓으로 여러 소켓과 데이터를 주고 받을 수 있으며, 데이터그램은 독립적으로 처리됩니다. 이러한 이유로 수신 측은 도착지 정보(DP, D-IP)만으로 Demultiplexing을 수행합니다.

TCP는 연결지향형 프로토콜, 즉 소켓과 소켓 사이에 1:1 연결을 맺습니다. 이 때문에 동일한 소켓으로 여러 소켓과 데이터를 주고 받을 수 없습니다. 따라서 각 연결을 명확히 구분하기 위해 도착지 정보 뿐만 아니라, 출발지 정보(SP, S-IP)도 필요합니다.

> ### Q3
> TCP 프로토콜에서 rdt1.0, 2.0, 3.0 각각에 대해 차이점을 기준으로 설명하라.

- rdt1.0
    - 하위 채널이 완전히 신뢰적이므로, 프로토콜은 단순히 데이터를 보내고 받는 역할 수행
- rdt2.0
    - 하위 채널에서 데이터 손실은 발생하지 않지만, 비트 오류가 발생 가능
    - 비트 오류를 검출하기 위해 송신 측에서 패킷에 체크섬 필드 추가
    - 수신 측에서는 패킷에 오류가 없다면 ACK를, 오류가 존재한다면 NAK를 송신 측에 전송
    - 수신 측에서 데이터 중복을 판단할 수 있도록 송신 측에서는 패킷에 순서 번호 필드 추가
    - 통신 과정을 간소화하기 위해 NAK 생략 가능
        
        이 경우 패킷에 오류가 존재할 때 수신 측에서는 NAK 대신 이전에 수신한 패킷에 대한 ACK 전송
        
- rdt3.0
    - 하위 채널에서 비트 오류 및 데이터 손실 발생 가능 (비신뢰적인 채널)
    - 비트 오류의 경우 rdt2.0과 동일하게 처리
    - 패킷을 전송했지만 하위 채널에서 손실되어 수신 측이 받지 못할 경우, 송신 측이 무한정 대기하는 것을 방지하기 위해 타이머 사용
    - 송신 측에서 패킷 전송 후 일정 시간이 지나도 ACK가 오지 않을 경우 패킷이 유실되었다고 판단하여 재전송

**Reply by Starlight258** [reply]
> ## Q1
> Multiplexing과 Demultiplexing의 개념에 대한 간단한 설명과  TCP와 UDP에서의 Multiplexing, Demultiplexing의 차이는 무엇인가?
> 
- 다중화란 애플리케이션 계층의 여러 소켓에서 전송되는 데이터를 하나로 모으는 것이다.
- 역다중화란 전송 계층 세그먼트의 데이터를 올바른 소켓으로 전달하는 것이다.

> 다중화는 데이터를 하나로 묶어 다음 계층으로 전달하는 것이고, 역다중화는 전달받은 데이터를 적절하게 분배하는 것이다.

- UDP 
  - 다중화: 소스 포트번호와 목적지 포트번호를 사용하여 세그먼트를 생성하며 TCP 연결을 수행한다.
  - 역다중화: 목적지 ip주소와 목적지 포트번호로 소켓을 식별하여 세그먼트를 전달한다.
> 같은 서버 소켓을 여러 클라이언트가 공유할 수 있다.

- TCP 
  - 다중화:  소스 포트번호와 목적지 포트번호를 사용하여 세그먼트를 생성하며 연결을 수행하지 않는다.
  - 역다중화: 소스 포트번호와 목적지 포트번호뿐만 아니라 소스 ip주소와 목적지 ip주소를 모두 사용하여 소켓을 식별하여 세그먼트를 전달한다.

> ## Q2
> DNS가 UDP를 사용하는 것 처럼, 일반적으로 네트워크 개발자는 TCP보다 UDP를 선호한다. 그 이유는 무엇인가?
> 
일반적으로 네트워크 개발자는 tcp를 udp보다 선호한다.
하지만 일부 애플리케이션은 tcp보다 udp가 적합한데, 아래와 같은 udp의 특징 때문이다.
- 데이터 전송 시점 제어 가능
  - udp는 tcp와 달리 혼잡제어를 하지 않으므로 데이터 전송이 지연되지 않는다.
- 연결 설정이 없음
  - udp는 데이터 전송 전에 연결을 수립하지 않아도 되므로 연결로 인한 지연이 없다.
- 연결 상태를 유지하지 않음
  - udp는 연결 상태를 유지하지않으므로 좀 더 많은 클라이언트를 수용할 수 있다.    
+ udp는 tcp에 비해 작은 패킷 오버헤드를 가진다.

> ## Q3
> TCP 프로토콜에서 rdt1.0, 2.0, 3.0 각각에 대해 차이점을 기준으로 설명하라.
### rdt1.0
- 완벽하게 신뢰적인 채널 상에서의 신뢰적인 데이터 전송

### rdt 2.0
- 비트 오류가 있는 채널 상에서의 신뢰적인 데이터 전송
- 전송 후 대기 프로토콜

### rdt 2.1
- ACK와 NAK가 손상되었을 경우를 대비하여 데이터 패킷에 순서번호를 추가하여 전송
- 순서가 바뀌었을 경우 : 중복 ACK
- 수신한 패킷이 손상되었을 경우 : NAK
> 중복 ACK : 손상된 패킷을 수신하면 가장 최근에 정확하게 수신한 패킷에 대해 ACK 송신

### rdt 2.2
- NAK 없이 ACK만 사용하는 신뢰적인 전송 프로토콜 
  - 순서가 바뀌거나 손상된 패킷이 수신되면 마지막으로 올바르게 수신한 패킷에 대한 순서번호를 포함하여 ACK를 전송한다. (중복 ACK)

### rdt 3.0 
- 비트 오류와 손실 있는 채널 상에서의 신뢰적 데이터 전송
- 하위 채널이 패킷을 손실하는 경우 송신자는 충분한 시간을 기다린 후 timeout시(ACK가 오지 않을 경우) 패킷을 재전송한다.

**Comment by Hanjaemo** [comment]
### Q1
도착지 정보(DP와 D-IP)로만 식별할 수 있는 UDP 소켓과 달리, TCP 소켓을 식별하려면 왜 출발지 정보(SP와 S-IP)도 함께 필요할까요?

---

### Q2
SR 프로토콜의 경우, 수신자가 현재 윈도우 base보다 작은 특정 순서 번호를 가진 이미 수신된 패킷을 무시하지 않고 재확인응답을 보내준다고 합니다. 이러한 이유는 무엇일까요?

---

### Q3
윈도우 크기는 왜 SR 프로토콜에 대한 순서 번호 공간 크기의 절반보다 작거나 같아야 할까요?

**Reply by H-sooyeon** [reply]
> **Q1 도착지 정보(DP와 D-IP)로만 식별할 수 있는 UDP 소켓과 달리, TCP 소켓을 식별하려면 왜 출발지 정보(SP와 S-IP)도 함께 필요할까요?**

이는 UDP와 TCP 프로토콜의 근본적인 차이로 인한 것이다.
UDP는 비연결형 프로토콜로, 데이터그램을 독립적으로 전송한다. 송신자와 수신자 간에 세션을 설정하지 않기 때문에 도착지 IP 주소와 도착지 포트 번호 DP만으로 소켓을 식별할 수 있다.
하지만, TCP는 연결형 프로토콜로, 데이터를 전송하기 전에 송신자와 수신자 간에 연결을 설정한다.
또한 양방향 통신을 지원하기 때문에, 동일한 도착지 IP와 포트로 여러 클라이언트가 접속할 수 있기 때문에, 클라이언트 각각을 구분하려면 출발지 IP와 포트가 필요하다.

<br />

> **Q2 SR 프로토콜의 경우, 수신자가 현재 윈도우 base보다 작은 특정 순서 번호를 가진 이미 수신된 패킷을 무시하지 않고 재확인응답을 보내준다고 합니다. 이러한 이유는 무엇일까요?**

이미 수신자가 패킷을 받았고 ACK도 보냈으나, ACK가 송신자에게 도달하지 못했다면 송신자는 불필요하게 그 패킷을 재전송하게 된다.
이때 수신자가 이미 받은 패킷에 대해 다시 ACK를 보내주면, 송신자는 그 패킷이 이미 안전하게 수신되었음을 확인할 수 있어 불필요한 재전송을 방지할 수 있다.
즉, Selective Repeat 프로토콜에서 이미 수신한 패킷에 대한 재확인응답을 보내는 이유는, 송신자와 수신자간의 통신을 신뢰성 있게 유지하고 불필요한 재전송을 줄이며, 네트워크 효율성을 높이기 위해서이다.

<br />

> **Q3 윈도우 크기는 왜 SR 프로토콜에 대한 순서 번호 공간 크기의 절반보다 작거나 같아야 할까요?**

송신자는 데이터를 연속적으로 전송하며, 각 패킷에는 고유한 순서 번호가 부여된다.
만약 윈도우 크기가 순서 번호 공간 크기의 절반보다 크다면, 수신자는 이미 수신한 패킷의 순서 번호와 새로 수신된 패킷의 순서 번호를 혼동할 수 있다.

ex) 순서 번호 공간이 0에서 7까지 있고, 윈도우 크기가 5라면, 송신자는 [0, 1, ,2, 3, 4]의 패킷을 전송한 후 [5, 6, 7, 0, 1]의 패킷을 보낼 수 있다. 이때 수신자가 [0, 1]을 다시 수신할 때 이전의 패킷인지, 새로운 패킷인지 구분할 수 없게 된다.

즉, 중복된 순서 번호의 패킷이 발생했을 때, 이를 정확히 식별하고 데이터 전송의 신뢰성과 순서를 보장하지 위해서이다.

**Reply by Starlight258** [reply]
> ### Q1
> 도착지 정보(DP와 D-IP)로만 식별할 수 있는 UDP 소켓과 달리, TCP 소켓을 식별하려면 왜 출발지 정보(SP와 S-IP)도 함께 필요할까요?
> 
- UDP는 연결을 설정하지 않으므로 여러 클라이언트가 같은 서버 소켓을 공유할 수 있다.
- TCP는 1:1 연결을 설정하므로 클라이언트별로 연결을 따로 설정하고 소켓도 따로 생성된다.
   연결된 각 소켓을 식별하기 위해 4개의 필드 모두를 사용한다.
> 한 출발지에서 목적지로 여러 소켓이 연결되어 있을 수 있다.

> ### Q2
> SR 프로토콜의 경우, 수신자가 현재 윈도우 base보다 작은 특정 순서 번호를 가진 이미 수신된 패킷을 무시하지 않고 재확인응답을 보내준다고 합니다. 이러한 이유는 무엇일까요?
> 
송신자와 수신자의 윈도우가 다를 수 있으므로, 동기화하기 위해서이다.
만약 ACK응답을 보내지 않는다면 송신자는 윈도우를 이동할 수 없다.

> ### Q3
> 윈도우 크기는 왜 SR 프로토콜에 대한 순서 번호 공간 크기의 절반보다 작거나 같아야 할까요?

윈도우 크기가 순서 번호 공간 크기의 절반보다 크다면, 수신자는 해당 패킷이 새로 전송된 패킷인지
재전송된 패킷인지 알 수 없다.

**Comment by Starlight258** [comment]
## Q1
**전송 계층**과 **네트워크 계층**은 어떤 역할을 하는지 비교해서 간단히 설명해주세요.
## Q2
**종단간의 원리**에 대해 간단히 설명해주세요.
## Q3 
**`GBN` 프로토콜**과 **`SR` 프로토콜**을 비교해서 간단하게 설명해주세요.

**Reply by H-sooyeon** [reply]
> Q1 전송 계층과 네트워크 계층은 어떤 역할을 하는지 비교해서 간단히 설명해주세요.

전송 계층은 종단 간 통신을 관리하고, 데이터 전송의 신뢰성을 보장한다.
데이터 분할 및 재조립, 흐름 제어, 오류 제어, 연결 설정 및 종료, 프로세스 간 통신 등의 기능을 수행하며 대표적으로 TCP, UDP 프로토콜이 존재한다.

네트워크 계층은 패킷을 목적지까지 전달하기 위한 경로를 선택하고, 패킷 전송을 관리한다.
라우팅, 패킷 포워딩, 논리적 주소 지정 등의 기능을 수행하며, 대표적으로 IP 프로토콜이 존재한다.

<br />

> Q2 종단간의 원리에 대해 간단히 설명해주세요.

네트워크의 주요 기능들은 네트워크의 중간 장치가 아닌 통신의 양 끝에 있는 종단 시스템에서 구현되어야 한다는 개념이다.
신뢰성 있는 데이터 전송, 오류 검출 및 복구, 데이터 암호화와 같은 중요한 통신 기능들은 네트워크의 중간 장치보다는 통신의 종단인 송신자와 수신자 시스템에서 수행되어야 한다.
즉, 네트워크 자체는 가능한 단순하게 설계하고, 복잡한 처리나 논리적인 기능은 종단 시스템에 맡기는 방식이다.
TCP 프로토콜로 예를 들면, 신뢰성 있는 데이터 전송을 보장하는 기능은 네트워크 중간 장치가 아닌 송신자와 수신자에서 구현된다.

<br />

> Q3 GBN 프로토콜과 SR 프로토콜을 비교해서 간단하게 설명해주세요.

둘 다 전송 계층에서 사용되는 오류 제어 매커니즘으로, 신뢰성 있는 데이터 전송을 보장하기 위한 방법을 제공한다.
Go-Back-N 프로토콜은 수신 측에서 어떤 패킷이 손실되거나 오류가 발생하면, 그 이후에 도착한 모든 패킷에 대해 ACK를 보내지 않는다. 송신 측은 타임아웃이 발생하면 손실된 패킷부터 그 이후의 모든 패킷을 다시 전송한다.

SR 프로토콜은 특정 패킷이 손실되거나 오류가 발생하면 송신 측은 해당 패킷만 재전송한다.
수신 측은 올바르게 수신된 패킷들을 버퍼링하고, 손실된 패킷이 재전송되면 모든 패킷을 올바른 순서로 재조립한다.

즉, GBN은 순서대로 수신하고 손실된 패킷 이후의 모든 패킷을 다시 전송해야 하는 단순한 방식이다.
SR은 순서에 관계없이 수신할 수 있으며, 손실된 패킷만 선택적으로 재전송하는 더 효율적이면서 복잡한 방식이다.

**Reply by Hanjaemo** [reply]
> ### Q1
> **전송 계층**과 **네트워크 계층**은 어떤 역할을 하는지 비교해서 간단히 설명해주세요.

<전송 계층>
- 서로 다른 호스트에서 동작하는 **프로세스들 간의 논리적인 통신을 제공**하는 계층
- 신뢰적인 데이터 전송, 혼잡 제어 등의 기능 제공
- 호스트 내에는 여러 프로세스들이 존재하며, 이를 구분하기 위해 포트 번호 사용

<네트워크 계층>
- **호스트들 간의 논리적인 통신을 제공**하는 계층
- 패킷을 목적지 IP 주소로 라우팅
- IP 주소를 사용하여 네트워크 내에서 호스트를 구분

> ### Q3
> **`GBN` 프로토콜**과 **`SR` 프로토콜**을 비교해서 간단하게 설명해주세요.

GBN 프로토콜과 SR 프로토콜은 모두 파이프라이닝 방식을 사용하여, 기존 전송 후 대기 방식의 성능을 향상시킨 프로토콜입니다. 

그러면 두 프로토콜은 어떤 부분에서 차이가 있을까요?

GBN 프로토콜의 경우, 패킷 하나의 오류 때문에 많은 패킷을 불필요하게 재전송한다는 문제점이 존재합니다. 
이는 곧 성능 이슈를 야기하므로, SR 프로토콜에서는 수신 윈도우 내의 순서 번호를 가진 패킷이 정상적으로 수신되는 경우, 수신 측에서는 다음과 같은 작업을 수행합니다.

1. 만약 범위에 속하는 패킷이 이전에 수신되지 않았던 것이라면 이를 버퍼에 저장합니다.
2. 만약 버퍼에 방금 저장한 패킷이 수신 윈도우의 base와 같은 순서 번호를 가졌다면, 해당 패킷과 이전에 버퍼에 저장되어 연속적인 번호를 가진 패킷들을 상위 계층으로 전달합니다.

**Reply by Starlight258** [reply]
> ## Q1
> **전송 계층**과 **네트워크 계층**은 어떤 역할을 하는지 비교해서 간단히 설명해주세요.
> 
- 전송 계층은 서로 다른 호스트에서 동작하는 `프로세스들 사이의 논리적 통신`을 제공한다.
- 네트워크 계층은 `호스트들 사이의 논리적 통신`을 제공한다.
> ## Q2
> **종단간의 원리**에 대해 간단히 설명해주세요.
> 
- `종단간의 원리`: 네트워크 기능은 가능한 한 통신의 끝단에서 구현되어야 한다.
- 상위 계층에서 서비스를 제공할때, 하위 레벨에 중복된 기능에 의존하여 구현할 수 없다. 하위 계층은 불확실성을 가지고 있기 때문이다.

> ## Q3
> **`GBN` 프로토콜**과 **`SR` 프로토콜**을 비교해서 간단하게 설명해주세요.
- `GBN 프로토콜` : 슬라이딩 윈도우를 기반으로 하며, 오류 발생시 수신자는 마지막으로 올바르게 수신한 패킷 이후의 모든 패킷을 재전송하도록 요청한다.
- `SR 프로토콜` : 슬라이딩 윈도우를 기반으로 하며, 오류 발생시 오류가 발생한 특정 패킷만을 선택적으로 재전송하도록 요청한다. 이를 통해 네트워크의 효율성을 높일 수 있다.
